rfcs/text/0007_lifecycle_unblocked.md

```diff
@@ -0,0 +1,290 @@
+3. Core should not expose API's as observables
+
+# Motivation
+Plugin lifecycle methods and context provider functions are async
+(promise-returning) functions. Core runs these functions in series and waits
+for each plugin's lifecycle function to resolve before calling the next. This
+allows plugins to depend on the API's returned from other plugins.
+
+With the current design, a single lifecycle method or context provider that
+blocks will block all of Kibana from starting up.
+
+We should make it impossible for a single plugin lifecycle function to
+stall all of kibana.
+
+# Detailed design
+
+### 1. Synchronous lifecycle methods
+Lifecycle methods are synchronous functions, they can perform asynchronous
+operations but Core doesn't wait for these to complete. This guarantees that
+no plugin lifecycle function can block other plugins or core from starting up.
+
+Core will still expose special API's that are able block the setup lifecycle
+such as registering Saved Object migrations, but this will be limited to
+operations where the risk of blocking all of kibana starting up is limited.
+
+### 2. Synchronous Context Provider functions
+Making context provider functions synchronous guarantees that a context
+handler will never be blocked by registered context providers. They can expose
+asynchronous API's which could potentially have blocking behaviour. 
+
+```ts
+export type IContextProvider<
+  TContext extends Record<string, any>,
+  TContextName extends keyof TContext,
+  TProviderParameters extends any[] = []
+> = (
+  context: Partial<TContext>,
+  ...rest: TProviderParameters
+) => TContext[TContextName];
+```
+
+### 3. Core should not expose API's as observables
+All Core API's should be reactive, when internal state changes their behaviour
+should change accordingly. But, exposing these internal state changes as part
+of the API contract leaks internal implementation details consumers can't do
+anything useful with and don't care about.
+
+For example: Core currently exposes `core.elasticsearch.adminClient$`, an
+Observable which emits a pre-configured elasticsearch client every time there's
+a configuration change. This includes changes such as the elasticsearch
+cluster `hosts` that alter the behaviour of the elasticsearch client. As a
+plugin author who wants to make search requests against elasticsearch I
+shouldn't have to care about, react to, or keep track of, how many times the
+underlying configuration has changed. I want to use the `callAsInternalUser`
+method and I expect Core to use the most up to date configuration to send this
+request to the correct `hosts`.
+
+This does not mean we should remove all observables from Core's API's. When an
+API consumer is interested in the *state changes itself* it absolutely makes
+sense to expose this as an Observable. Good examples of this is exposing
+plugin config as this is state that changes over time to which a plugin should
+directly react to.
```

Comment:
With this change, how would a plugin (or Core, even) handle these things:
- When the Elasticsearch config changes (to an entirely different cluster, for example) I need to make sure an index exists for my plugin. If the clients hide that change, I don't have a way to react to it as a plugin author. Now my plugin is in a broken state that can only be solved by restarting Kibana. You _could_ fix this by checking that setup logic has been executed in a pre-flight check, but that seems much more complicated than just being able to react to changes.
- Imagine my plugin is in the middle of executing some business logic that depends on data I queried from ES and then writes an update back to ES (multiple API calls). What happens when the ES cluster changes between ES API calls? Now I'm in a potentially inconsistent state and there's really no way for my plugin to handle such a situation (or even know it happened).

It's possible we could solve this problem by ensuring that the underlying client doesn't change in a context handler, but we already do that. It seems like this proposal would really have to be: don't expose ES clients in CoreSetup or CoreStart at all, only expose via context handlers (which are expected to complete eventually?).

This kind of just feels like moving the problem from one place to another, but maybe that's what we want in order to support a more stable Kibana experience.

Reply:
Thanks, these are important scenarios the proposal doesn't really address. Reactivity introduces complexity and I agree this is really just about where we move this "problem".

If we implement (1) and (2) but not (3) we move more of this complexity into plugins that want to expose API's and into their context handlers.

My assumption has been that plugins don't care about or use the information that a elasticsearch client has changed, but if they need to recover from the above scenarios we have to expose this to them and then it justifies the added complexity.
> This includes changes such as the elasticsearch cluster `hosts` that alter the behaviour of the elasticsearch client. As a plugin author who wants to make search requests against elasticsearch I shouldn't have to care about, react to, or keep track of, how many times the underlying configuration has changed.

I think it makes sense to explicitly say which kinds of changes we expect plugins to recover from seamlessly / transparently and for which changes is it OK to simply go into an unrecoverable error state forcing a kibana restart.

Unless we know that all plugins can successfully recover from a change in elasticsearch cluster, it might be better to force a sysadmin to restart kibana instead of saying "here's the mechanism to change your cluster host, but we really don't know what will happen if you use it, you might have to restart kibana anyway."

Reply:
FWIW, I think it's appropriate for clients to be consistent within a single context, and I also think it's appropriate (even ideal) to not pass clients to setup/start.

This is definitely moving the problem to a different place, but it's a more sensible place. Context is short lived whereas plugin lifecycle is long-lived. We're basically considering two options: block all of Kibana or block certain features of Kibana. When do we legitimately want to block all of Kibana?

Real world example: Why would all of Kibana, including dev tools and the plethora of stack monitoring and management tools be unavailable just because the task manager index failed to create?

The ES cluster changing today is a problem that exists pretty much exclusively due to log rotation. The legacy ES client configures its logger via its constructor, so when we rotate log config in Kibana we need to construct a new ES client. The new client doesn't have this problem, but until we switch to it we need to accommodate the client swapping.

Reply:
OK, so conceptually we could whitelist some configuration options to which Core will automatically react like logging or https://github.com/elastic/kibana/issues/19829. These options would change the internals of the Elasticsearch client but there shouldn't be a need for consumers to be aware of these underlying changes - it can be hidden under an API facade.

Ideally, plugins should be resilient when there are major changes like a snapshot restore dropping indexes, but these scenarios won't be _caused_ by Core so there would be no way to detect and no need to expose this from Core's Elasticsearch client.

Reply:
I believe this has been resolved by the note added above about not changing the underlying Elasticsearch client for some configuration changes.